File: extract_pdf.mjs
import fs from 'fs';
import pdfParse from 'pdf-parse';

function cleanText(text) {
    const lines = text.split('\n');
    const cleaned = [];
    for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed) {
            if (cleaned.length > 0 && !/[.!?:]$/.test(cleaned[cleaned.length - 1])) {
                cleaned[cleaned.length - 1] += ' ' + trimmed;
            } else {
                cleaned.push(trimmed);
            }
        }
    }
    return cleaned.join('\n');
}

function parseDomain(text) {
    const match = text.match(/Domain (\d+): (.+)/);
    if (match) {
        return {
            number: parseInt(match[1]),
            name: match[2].trim(),
            label: `Domain ${match[1]}: ${match[2].trim()}`
        };
    }
    return null;
}

function parseAspek(text) {
    const match = text.match(/Aspek (\d+): (.+)/);
    if (match) {
        return {
            number: parseInt(match[1]),
            name: match[2].trim(),
            label: `Aspek ${match[1]}: ${match[2].trim()}`
        };
    }
    return null;
}

function parseIndikator(text) {
    const match = text.match(/Indikator (\d+): (.+)/);
    if (match) {
        return {
            number: parseInt(match[1]),
            name: match[2].trim(),
            label: `Indikator ${match[1]}: ${match[2].trim()}`
        };
    }
    return null;
}

function extractSections(text) {
    const domainPattern = /(?=Domain \d+:)/g;
    const sections = text.split(domainPattern).filter(s => s.trim());

    const data = [];
    let currentDomain = null;
    let currentAspek = null;

    for (const section of sections) {
        const domain = parseDomain(section);
        if (domain) {
            currentDomain = domain;
            const aspekMatch = section.match(/Aspek \d+: .+?(?=Indikator \d+:|$)/s);
            if (aspekMatch) {
                currentAspek = parseAspek(aspekMatch[0]);
            } else {
                currentAspek = null;
            }
        } else {
            continue;
        }

        const indikatorPattern = /Indikator \d+: .+?(?=Indikator \d+:|Aspek \d+:|Domain \d+:|$)/gs;
        const indikators = section.match(indikatorPattern) || [];

        for (const indText of indikators) {
            const indikator = parseIndikator(indText);
            if (indikator) {
                const contentMatch = indText.match(/Indikator \d+: .+?\n(.+)/s);
                const content = contentMatch ? contentMatch[1].trim() : '';
                data.push({
                    metadata: {
                        domain: currentDomain,
                        aspek: currentAspek,
                        indikator: indikator
                    },
                    content: content
                });
            }
        }
    }

    return data;
}

const dataBuffer = fs.readFileSync('public/sample.pdf');

pdfParse(dataBuffer).then(function(data) {
    let extractedText = '';
    data.pages.forEach((page, index) => {
        if (index >= 29 && index <= 76) { // Pages 30-77 (0-indexed 29-76)
            extractedText += page.text + '\n';
        }
    });

    const cleanedText = cleanText(extractedText);
    const sectionsData = extractSections(cleanedText);

    fs.writeFileSync('extracted_data.json', JSON.stringify(sectionsData, null, 4), 'utf8');
    console.log('Extraction complete. Data saved to extracted_data.json');
}).catch(function(error) {
    console.error('Error parsing PDF:', error);
});

File: extract_pdf.py
import pdfplumber
import re
import json

def clean_text(text):
    lines = text.split('\n')
    cleaned = []
    for line in lines:
        line = line.strip()
        if line:
            if cleaned and not re.search(r'[.!?:]$', cleaned[-1]):
                cleaned[-1] += ' ' + line
            else:
                cleaned.append(line)
    return '\n'.join(cleaned)

def parse_domain(text):
    match = re.search(r'Domain (\d+): (.+)', text)
    if match:
        return {
            "number": int(match.group(1)),
            "name": match.group(2).strip(),
            "label": f"Domain {match.group(1)}: {match.group(2).strip()}"
        }
    return None

def parse_aspek(text):
    match = re.search(r'Aspek (\d+): (.+)', text)
    if match:
        return {
            "number": int(match.group(1)),
            "name": match.group(2).strip(),
            "label": f"Aspek {match.group(1)}: {match.group(2).strip()}"
        }
    return None

def parse_indikator(text):
    match = re.search(r'Indikator (\d+): (.+)', text)
    if match:
        return {
            "number": int(match.group(1)),
            "name": match.group(2).strip(),
            "label": f"Indikator {match.group(1)}: {match.group(2).strip()}"
        }
    return None

def extract_sections(text):
    # Split text into sections based on Domain
    domain_pattern = r'(?=Domain \d+:)'
    sections = re.split(domain_pattern, text)
    sections = [s.strip() for s in sections if s.strip()]

    data = []
    current_domain = None
    current_aspek = None

    for section in sections:
        domain = parse_domain(section)
        if domain:
            current_domain = domain
            # Find aspek in this section
            aspek_match = re.search(r'Aspek \d+: .+?(?=Indikator \d+:|$)', section, re.DOTALL)
            if aspek_match:
                current_aspek = parse_aspek(aspek_match.group(0))
            else:
                current_aspek = None
        else:
            # If no domain, perhaps continuation or error
            continue

        # Find all indikator in this section
        indikator_pattern = r'Indikator \d+: .+?(?=Indikator \d+:|Aspek \d+:|Domain \d+:|$)'
        indikators = re.findall(indikator_pattern, section, re.DOTALL)

        for ind_text in indikators:
            indikator = parse_indikator(ind_text)
            if indikator:
                # Extract content after the indikator label
                content_match = re.search(r'Indikator \d+: .+?\n(.+)', ind_text, re.DOTALL)
                content = content_match.group(1).strip() if content_match else ""
                data.append({
                    "metadata": {
                        "domain": current_domain,
                        "aspek": current_aspek,
                        "indikator": indikator
                    },
                    "content": content
                })

    return data

# Main
with pdfplumber.open('public/sample.pdf') as pdf:
    text = ''
    for page_num in range(29, 77):  # Pages 30-77 (0-indexed 29-76)
        if page_num < len(pdf.pages):
            page = pdf.pages[page_num]
            page_text = page.extract_text()
            if page_text:
                text += page_text + '\n'

cleaned_text = clean_text(text)
sections_data = extract_sections(cleaned_text)

with open('extracted_data.json', 'w', encoding='utf-8') as f:
    json.dump(sections_data, f, ensure_ascii=False, indent=4)

print("Extraction complete. Data saved to extracted_data.json")